compose/ui/ui-uikit/src/uikitMain/objc/CMPUIKitUtils/CMPUIKitUtils/CMPViewController.m

```diff
@@ -52,176 +25,101 @@ @interface UIViewController(CMPUIKitUtilsPrivate)
+    
+    if (parent) {
+        [self transitLifecycleToStarted];
     }
-
-    [self cmp_setNeedsVerifyHierarchy];
-}
-
-- (void)removeFromParentViewController {
-    [super removeFromParentViewController];
-
-    [self cmp_setNeedsVerifyHierarchy];
 }
 
-- (void)viewDidLoad {
-    if ([self.view isKindOfClass:[CMPView class]]) {
-        __weak typeof(self) weakSelf = self;
-        ((CMPView *)self.view).windowDidChangeBlock = ^{
-            [weakSelf cmp_setNeedsVerifyHierarchy];
-        };
-    } else {
-        NSAssert(false, @"The view of CMPViewController should be either CMPView or subclassed from CMPView");
+/// This method can be triggered by both `viewWillAppear` and `didMoveToParentViewController`. Mind that in second case the `view` could still not be constructed (adding `UIViewController`  to inactive tab of `UITabBarController`, for example), so the `ComposeScene` is not neccessarily created at this point.
+- (void)transitLifecycleToStarted {
+    switch (_lifecycleState) {
+        case CMPViewControllerLifecycleStateDestroyed:
+            @throw [NSException exceptionWithName:@"CMPViewControllerMisuse"
+                                           reason:@"CMPViewController shouldn't be reused after completely removed from hierarchy, because it's logically marked as Destroyed. You must create a new CMPViewController and use it instead."
+                                         userInfo:nil];
+        case CMPViewControllerLifecycleStateInitalized:
+            _lifecycleState = CMPViewControllerLifecycleStateStarted;
+            [self viewControllerDidEnterWindowHierarchy];
+            [self scheduleHierarchyContainmentCheck];
+            break;
+        case CMPViewControllerLifecycleStateStarted:
+            break;
     }
-
-    [super viewDidLoad];
-}
-
-- (void)loadView {
-    self.view = [CMPView new];
 }
 
-- (void)cmp_setNeedsVerifyHierarchy {
-    if (self.cmp_needsVerifyHierarchy) {
-        return;
-    }
-    self.cmp_needsVerifyHierarchy = true;
-
-    __weak typeof(self) weakSelf = self;
-    dispatch_async(dispatch_get_main_queue(), ^{
-        weakSelf.cmp_needsVerifyHierarchy = false;
-        [weakSelf cmp_verifyHierarchyAndScheduleNotificationIfNeeded];
+- (void)scheduleHierarchyContainmentCheck {
+    double delayInSeconds = 0.5;
+    
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(delayInSeconds * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{        
+        switch (self->_lifecycleState) {
+            case CMPViewControllerLifecycleStateInitalized:
+            case CMPViewControllerLifecycleStateDestroyed:
+                assert(false);
+                break;
+            case CMPViewControllerLifecycleStateStarted:
+                // perform check
+                if ([self cmp_isInWindowHierarchy]) {
+                    // everything is fine, schedule next one
+                    [self scheduleHierarchyContainmentCheck];
```

Comment:
I still a bit worry about random blinks of `[self cmp_isInWindowHierarchy]` result in a various scenarios.
Can we add a second check after a single run loop cycle one more time before destroying compose scene?